src/box/relay.cc

```diff
@@ -642,7 +643,15 @@ static void
 tx_gc_advance(struct cmsg *msg)
 {
 	struct relay_gc_msg *m = (struct relay_gc_msg *)msg;
-	gc_consumer_advance(m->relay->replica->gc, &m->vclock);
+	struct credentials *old_user = effective_user();
+
+	fiber_set_user(fiber(), &admin_credentials);
+	if (box_gc_consumer_update(m->relay->replica, &m->vclock) != 0) {
+		say_error("Cannot advance gc consumer for replica %u",
+			  m->relay->replica->id);
+		diag_log();
+	}
+	fiber_set_user(fiber(), old_user);
```

Comment:
On subscribe, we should advance gc consumer synchronously.

Consider the case:
Replica just joined the cluster, it has actual consumer on master and dummy consumers on replicas (they don't pin any xlogs). When it subscribes to another replica, it should start pinning needed xlogs before it started to send data - that's why we need a synchronous write there.

If we updated gc consumer asynchronously (or create a job for GC fiber), there would a moment when replica is subscribed and receives xlogs, but needed xlogs are not pinned and can be deleted.

To relieve this pain, we could set in-memory consumer along with creating a job for GC fiber on subscribe, but then xlogs can be pruned after restart if the asynchronous write wasn't completed. In this case, our GC is not truly persistent. Also, update of persistent consumer can fail - it makes the problem worse (even if we had retries for it). 

So `TXN_FORCE_ASYNC` is still needed, but after your commit it will be moved to `box_process_subscribe`.

Reply:
> Have you thought about having a special system fiber in gc subsystem for writing these GC advancements?

Thanks, that's a great idea! I've been thinking about it.

If we want GC to be truly persistent, we have to synchronously update consumers in those places:
- JOIN - right after replica registration, so the master started pinning xlogs before final join
- SUBSCRIBE - we should pin xlogs before sending rows
- REGISTER - (related to EE part) when anonymous replica is registered, it can FETCH_SNAPSHOT right after registration, and its dummy consumer must be created before REGISTER is over
- FETCH_SNAPSHOT - (related to EE part) after an anonymous replica fetched the snapshot, we should update its consumer: set `with_snap` flag to `false` so that we can check in SUBSCRIBE that the replica really fetched the whole snapshot. Also, I believe that we should synchronously update consumer in the beginning of FETCH_SNAPSHOT and send rows from snap only after the consumer is persisted and the snap is pinned.

We can asynchronously update consumers in these cases:
- JOIN - after final join
- SUBSCRIBE - in `tx_gc_advance` message from relay (when xlog is closed)

The latter (`tx_gc_advance`) should be (or even must be) asynchronous. That's not so in my patch, so, AFAIU, workers of "tx" can become littered with my updates in the case of long WAL write or waiting for synchronous transactions to be confirmed.

So we could create a special fiber for these two cases, but TXN_FORCE_ASYNC won't go away from subscribe and the fiber won't encapsulate whether the consumers are persistent anyway, so the only problem that would be solved is non-yielding `gc_consumer_advance`. But I can simply create a function `gc_consumer_advance_async` that will call `txn_commit_try_async` instead of `txn_commit` and then the yield will go away from `tx_gc_advance`. But on the other hand, if we created a system GC fiber, we could implement retries for failed updates (we weren't planning on doing them, but why not). What do you think about it?

Reply:
> we have to synchronously update consumers in those places

This wouldn't be a problem. You could add a function `gc_consumer_advance_now(consumer, timeout)`, blocking, so it would have to wait until the update is persisted. Via `fiber_cond_wait()` internally, for example.

Reply:
> So we could create a special fiber for these two cases

A special fiber isn't just for async work though. It would be a central source of truth. Consider this: your GC consumers would have a volatile vclock updated by each async gc-advance, and a persisted vclock, updated only on commit.

GC consumers are not re-created on each advance. Instead, they update their volatile vclock right away, set a flag "is_volatile", wakeup the worker fiber, and it finds all "volatile" consumers, and tries to persist whatever they have right now in the volatile vclock.

This is similar to how core/raft is working with terms and votes. It applies the volatile state right away, but the persisted one is catching up in a worker fiber.

If persist fails, the worker can retry later. If persist succeeds, the worker will update the GC's persisted vclock and proceed with deleting files if needed. According to the persisted vclock.

Note that the volatile vclock isn't updated on-replace. It has to be in-code change by us. If somebody comes and manually changes `_gc_consumer`, we can on-commit update the volatile state to it too if it was equal to the old non-volatile state.

Also note that we can't drop the volatile vclock. It is needed for relay to advance GC consumer asynchronously. If all usecases would be blocking, we wouldn't need volatile and could just update stuff always on-commit and store only the persisted vclock.

Right now you are trying to solve it in a decentralized way that all consumers are on their own. But perhaps a centralized solution would be simpler?

Reply:
Unrelated to worker fiber vs always blocking: it seems we could make all GC consumer txns force-async, right? Is there any reason why some of them have to comply with the limbo? If not, then we could encapsulate the force-async flag inside gc.c and just use it for all consumers. Or?

I see that you try to update the consumer in a txn with `_cluster` update here `box_register_replica_with_consumer()`. But you probably don't need that? Why does it have to be one txn?

Reply:
> so the only problem that would be solved is non-yielding gc_consumer_advance

No, we would also have one manager of all these consumers + ability to retry. You wouldn't need to re-create the consumers anymore. You just update the volatile vclock and wakeup the worker.

Reply:
After re-reading your other comment about the on-commit vs on-replace update I feel like you could get away with just switching to on-commit and not having any GC fibers. The problem, which I was trying to solve with on-replace update using the fiber, wouldn't exist, if you would update on-commit.

We wouldn't have async advance for relay though. For them you would need to make an additional worker fiber per-relay which would be doing what I explained above for all GC consumers, but just for this relay. I.e. a volatile GC vclock, and the worker persisting it when it sees the vclock changed.

`txn_commit_try_async()` isn't exactly our thing here, because it is only an attempt of async update. It blocks when WAL queue is full. Any sort of "always async push a txn into a queue and move on" potentially leads to that queue overflowing. That is why whatever we decide, it would either be blocking, or
"eventual consistency" when you update the volatile state, and make the best
effort in an async worker to save it into WAL.

A side note - `tx_status_update()` already yields in `txn_limbo_ack()`, but that isn't good, has to be changed, and we shouldn't make it worse here.

Reply:
To summarize:

### Consumer = tuple vs consumer = state

Both `one-tuple=one-gc-consumer` solution and `gc-consumer=state` solution work. Still, I suggest to implement the second option and see if it is simpler. Can't tell right away, but I feel it could actually be simpler.

### Async globally or async in relay

You need to make relay advancement async. Unavoidably it would be a worker fiber. Either per-relay and relatively simple, or one for the whole GC subsystem.

Unfortunately, I can't tell what would look simpler in total. I recommend to implement both and compare. And ask Vova. He probably won't come here or when he does, he would suggest something new and much simpler but too late. Better ping him in a chat or in person.

Reply:
I significantly redesigned the patch, PTAL.
I encapsulated gc consumers machinery in gc subsystem and chosen a global fiber for async updates of consumers.
However, I failed to encapsulate all the "truth" in one fiber - we register and unregister consumers from on_replace trigger of `_cluster`, hence, inside active transaction, so we cannot yield and wait for other fiber to complete.

Reply:
>  it seems we could make all GC consumer txns force-async

Only updates (and I did so in redesigned patch). What's about [un]register, we register and unregister consumers from on_replace trigger of `_cluster`, so we cannot set this flag there.